src/data/finsupp/basic.lean

```diff
@@ -1717,6 +1717,13 @@ lemma sum_smul_index' [semiring R] [add_comm_monoid M] [semimodule R M] [add_com
   (b • g).sum h = g.sum (λi c, h i (b • c)) :=
 finsupp.sum_map_range_index h0
 
+@[simp]
+lemma smul_eq_zero [semiring R] [no_zero_divisors R] {c : R} {f : ι →₀ R} :
+  c • f = 0 ↔ c = 0 ∨ f = 0 :=
+finsupp.ext_iff.trans
+  ⟨λ h, or_iff_not_imp_left.mpr (λ hc, finsupp.ext (λ i, (mul_eq_zero.mp (h i)).resolve_left hc)),
+   λ h, by { cases h; simp [h] }⟩
+
```

Comment:
I would really like a typeclass for this!

Reply:
You mean something like:
```
class no_zero_smul_divisors R M := (smul_eq_zero : ∀ c x, c • x = 0 ↔ c = 0 ∨ x = 0)
```

Reply:
Yes, exactly - although maybe the implication only needs to go one way

Reply:
#6199